### PR TITLE
Video process complete lambda

### PR DESCRIPTION
**Pull request by angellee-io**
Missing push notification and update thumbnail_id (TBD)

### CODE REVIEW

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	public VideoProcessCompletedRequestHandler() {
+		databaseManager = DatabaseManager.getInstance();
+	}
+
+	public VideoProcessCompletedRequestHandler(DatabaseManager manager) {
+		this.databaseManager = manager;
+	}
+
+	@Override
+	public Response handleRequest(S3Event input, Context context) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: How come this is an S3 event? I thought this was listening on SNS event?

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+		if (updateSuccessful) {
+			LOG.info("updateToProcessCompleted");
+			// Send push notications
+		}
+				
+		return new Response(updateSuccessful);
+	}
+
+	private String getFileIdFromS3Record(S3EventNotificationRecord record) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: Is this the exact same function as the one inside VideoUploadCompletedRequestHandler.java ?

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+		String[] segments = s3Key.split("/");
+		return segments[1];
+	}
+
+	private Boolean updateToProcessCompleted(String fileId) {
+		GetVideoByFileIdQuery getQuery = new GetVideoByFileIdQuery(fileId);
+		Video videoInDatabase = databaseManager.execute(getQuery);
+		videoInDatabase.setWasProcessingCompleted(true);
+
+		UpdateVideoQuery updateQuery = new UpdateVideoQuery(videoInDatabase);
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: You don't need a whole UpdateVideoQuery. manager.update(videoInDatabase) is the same thing.

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java (hunk 1 above)
Comment (angellee-io) [reply src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: No this is an s3 event that gets triggered on master-playlist.m3u8 creation. So it's a S3Event.
Another way to do this is to setup job completion notification on transcoder pipeline and that will involve sns.

Comment (aaljebouri) [reply src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: Wouldn't doing it from the job notification allow you to do this and the thumbnail in 1 lambda instead of 2?

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java (hunk 2 above)
Comment (angellee-io) [reply src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: I believe they are just implemented a little differently.

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java (hunk 3 above)
Comment (angellee-io) [reply src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: But I want the rollback ability

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java (hunk 2 above)
Comment (aaljebouri) [reply src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: Merge them?

File: src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java (hunk 3 above)
Comment (aaljebouri) [reply src/com/grideo/videoProcessing/requestHandlers/VideoProcessCompletedRequestHandler.java]: The rollback ability is only useful if you are executing multiple sql statements. Otherwise there is not much point in transactions. Your query will either work or fail.

File: src/com/grideo/videoProcessing/notifications/PushNotificationManager.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+			instance = new PushNotificationManager(environmentManager);
+		}
+
+		stopwatch.stop();
+		LOG.info("Duration PushNotificationManager.getInstance(): " + stopwatch.getTime() + "ms");
+
+		return instance;
+	}
+	
+	public String registerUserForPush(String customUserData, String token, EndpointType endpointType) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/PushNotificationManager.java]: Why is this here? Is anyone going to call this lambda to register from push?

File: src/com/grideo/videoProcessing/notifications/PushNotificationManager.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		} else {
+			request.setPlatformApplicationArn(applePushApplication);
+		}
+		request.setToken(token);
+		request.setCustomUserData(customUserData);
+		CreatePlatformEndpointResult result = snsClient.createPlatformEndpoint(request);
+		return result.getEndpointArn();
+	}
+
+	public void unregisterUserFromPush(String endpointArn) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/PushNotificationManager.java]: Same.

File: src/com/grideo/videoProcessing/notifications/PushNotificationManager.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		DeleteEndpointRequest request = new DeleteEndpointRequest();
+		request.setEndpointArn(endpointArn);
+		snsClient.deleteEndpoint(request);
+	}
+
+	private AmazonSNS buildSnsClient(Regions region) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/PushNotificationManager.java]: Same.

File: src/com/grideo/videoProcessing/notifications/PushNotificationManager.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+		AmazonSNS amazonSns = AmazonSNSClient.builder().withRegion(region).build();
+
+		stopwatch.stop();
+		LOG.info("Duration NotificationManager.AmazonSNS: " + stopwatch.getTime() + "ms");
+
+		return amazonSns;
+	}
+
+	private String getSnsApplicationArnKey(
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/PushNotificationManager.java]: Same.

File: src/com/grideo/videoProcessing/notifications/PushNotificationManager.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+			String systemEnvSnsApplicationArnKey= systemEnvironmentVariables.get(EnvironmentManager.SNS_APPLICATION_ARN_KEY);
+			if (systemEnvSnsApplicationArnKey != null) {
+				snsApplicationArnKey = parametersResultMap.get(systemEnvSnsApplicationArnKey);
+			}
+		}
+
+		return snsApplicationArnKey;
+	}
+	
+	private String getAndroidSnsApplicationArnKey(
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/PushNotificationManager.java]: Same.

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    public String getEncoderFunction() {
+    	return encoderFunction;
+    }
+    
+    private void populateSystemEnvironmentVariables() {
+        Map<String,String> environmentVariables = new HashMap<>();
+
+        environmentVariables.put(IS_DEBUG, System.getenv(IS_DEBUG));
+        environmentVariables.put(AWS_DEFAULT_REGION, System.getenv(AWS_DEFAULT_REGION));
+        environmentVariables.put(SNS_APPLICATION_ARN_KEY, System.getenv(SNS_APPLICATION_ARN_KEY));
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: Needed?

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+import java.util.*;
+
+public class EnvironmentManager {
+
+    public static final String IS_DEBUG = "IsDebug";
+    public static final String AWS_DEFAULT_REGION = "AWS_DEFAULT_REGION";
+    public static final String NOTIFICATION_FUNCTION = "NotificationFunction";
+    public static final String ENCODER_FUNCTION = "EncoderFunction";
+    public static final String SNS_APPLICATION_ARN_KEY = "SNSApplicationArnKey";
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: Needed?

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+    
+    private EnvironmentManager() {
+        populateSystemEnvironmentVariables();
+
+        if (systemEnvironmentVariables.get(IS_DEBUG) == null ||
+                systemEnvironmentVariables.get(IS_DEBUG).equals(Boolean.FALSE.toString())) {
+            populateRegionFromSystemEnvironmentVariables(systemEnvironmentVariables);
+            populatePushFromSystemEnvironmentVariables(systemEnvironmentVariables);
+            populateEncoderFromSystemEnvironmentVariables(systemEnvironmentVariables);
+            GetParametersRequest parametersRequest = buildGetParametersRequest(systemEnvironmentVariables);
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: I don't think there is anything in this lambda that we need to go to parameters store for.

File: src/com/grideo/videoProcessing/notifications/NotificationManager.java [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+		List<PushNotificationRequest> notifications = new ArrayList<>();
+		for (Endpoint endpoint : receiver.getEndpoints()) {
+			String message = NotificationStringGenerator.generateStringForShareVideoPush(senderFirstName);
+			notifications.add(generatePushMessage(endpoint, message));
+		}			
+		
+		return notifications;
+	}
+	
+	public List<NotificationRequest> getPushNotificationsForCommentRecipient(User receiver, String senderName, String videoTitle) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/NotificationManager.java]: Needed?

File: src/com/grideo/videoProcessing/notifications/NotificationManager.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+		List<NotificationRequest> notifications = new ArrayList<>();
+		for (Endpoint endpoint : receiver.getEndpoints()) {
+			String message = NotificationStringGenerator.generateStringForCreateCommentPush(senderName, videoTitle, groupName);
+			notifications.add(generatePushMessage(endpoint, message));
+		}			
+		
+		return notifications;
+	}
+		
+	public List<NotificationRequest> getPushNotificationsForAddToGroupRecipient(User receiver, String senderName, String groupName) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/NotificationManager.java]: Needed?

File: src/com/grideo/videoProcessing/notifications/NotificationManager.java [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+		for (Endpoint endpoint : receiver.getEndpoints()) {
+			String message = NotificationStringGenerator
+					.generateStringForAddToGroupPush(senderName, groupName);
+			notifications.add(generatePushMessage(endpoint, message));
+		}			
+		
+		return notifications;
+	}	
+	
+	public List<NotificationRequest> getPushNotificationsForAddContact(User receiver, String senderName) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/notifications/NotificationManager.java]: Etc. for the rest of the file?

File: src/com/grideo/videoProcessing/sql/videos/queries/UpdateVideoQuery.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
+package com.grideo.videoProcessing.sql.videos.queries;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.hibernate.Session;
+
+import com.grideo.videoProcessing.sql.IQuery;
+import com.grideo.videoProcessing.sql.videos.Video;
+
+public class UpdateVideoQuery implements IQuery<Boolean> {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/sql/videos/queries/UpdateVideoQuery.java]: This query is not needed.

**Comment by angellee-io** [comment]
What email?

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java (hunk 9 above)
Comment (angellee-io) [reply src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: Aren't these needed to handle different environment?

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java (hunk 11 above)
Comment (angellee-io) [reply src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: the SNS arns?

**Comment by aaljebouri** [comment]
screen shot 2017-09-20 at 12 59 24 pm

grideoVideoInvite email.

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java (hunk 11 above)
Comment (aaljebouri) [reply src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: Why do you need the SNS arns if this function isn't registering/unregistering people to push?

File: src/com/grideo/videoProcessing/environment/EnvironmentManager.java (hunk 9 above)
Comment (aaljebouri) [reply src/com/grideo/videoProcessing/environment/EnvironmentManager.java]: But this function isn't doing the endpoint registration, why is it needed?

**Comment by angellee-io** [comment]
I still don't know what you mean.. what grideoVideoInvite?

**Comment by aaljebouri** [comment]
Did you test sending a video to a user that is not in the system and them getting an email when the video processing completes? You need to include grideoVideoInvite.html otherwise how will the email be generated?

**Comment by angellee-io** [comment]
then what's the jira tix for?

**Comment by aaljebouri** [comment]
The Jira ticket is because of the duplicated code, I feel like we need a better way to do it, but for now this way works. The ticket is for us to go back after release and clean up.

**Comment by angellee-io** [comment]
https://aequilibrium.atlassian.net/browse/GRID30-872

File: src/com/grideo/videoProcessing/strings/SMSStringGenerator.java [hunk 16]
```diff
@@ -0,0 +1,18 @@
+package com.grideo.videoProcessing.strings;
+
+public class SMSStringGenerator {
+	public static String generateStringForAddContact(String senderFullName, String recipientPhoneNumber) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/strings/SMSStringGenerator.java]: Not needed.

File: src/com/grideo/videoProcessing/strings/SMSStringGenerator.java [hunk 17]
```diff
@@ -0,0 +1,18 @@
...
+		return senderFullName + " has invited you to join Grideo. To connect with them download the Grideo app and use " +
+				recipientPhoneNumber + " when you Sign Up! http://www.grideo.com";
+	}
+
+	public static String generateStringForAddToGroup(String senderFullName, String groupName, String recipientPhoneNumber) {
```
Comment (aaljebouri) [inline src/com/grideo/videoProcessing/strings/SMSStringGenerator.java]: Not needed.
